### PR TITLE
KEP-4222: Make CBOR roundtrip tests pass through interface{}.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind cleanup
/sig api-machinery

#### What this PR does / why we need it:

Adds coverage for decode-to-interface{} and encode-from-interface{} paths to the CBOR roundtrip tests, as requested on https://github.com/kubernetes/kubernetes/pull/123267.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/4222
```

### CODE REVIEW

**Comment by benluddy** [comment]
/assign @deads2k

This is the follow-up to your feedback on https://github.com/kubernetes/kubernetes/pull/123267.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go [hunk 1]
```diff
@@ -101,10 +102,6 @@ func TestRoundtrip(t *testing.T) {
 			name: "int64 zero",
 			obj:  int64(math.MinInt64),
 		},
-		{
```
Comment (benluddy) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go:104]: Added this case to the tracking issue. It's not roundtrippable with JSON, either, but it does fail differently. Related to the tracked TODO about int64 overflow behavior.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go [hunk 2]
```diff
@@ -105,13 +105,13 @@ func TestDecode(t *testing.T) {
 			assertOnError: assertNilError,
 		},
 	} {
-		ms := tc.modes
-		if len(ms) == 0 {
-			ms = allDecModes
+		decModes := tc.modes
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go:108]: just can't bring yourself to go to decodeModes? :)

Comment (benluddy) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go:108]: Hold on, are you telling me "dec" is an abbreviation of "decode"?

Comment (benluddy) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go:108]: (I went with the type name https://pkg.go.dev/github.com/fxamacker/cbor/v2#DecMode here).

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go [hunk 3]
```diff
@@ -260,40 +257,50 @@ func TestRoundtrip(t *testing.T) {
...
+			decModeName, ok := decModeNames[modePair.dec]
 			if !ok {
 				t.Fatal("test case configured to run against unrecognized decode mode")
 			}
 
 			t.Run(fmt.Sprintf("enc=%s/dec=%s/%s", encModeName, decModeName, tc.name), func(t *testing.T) {
 				original := tc.obj
 
-				b, err := mp.enc.Marshal(original)
+				b, err := modePair.enc.Marshal(original)
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go:285]: Please *also* check that these bytes are unmarshalable into the final.Interface().

I think we currently have

1. typedObject to typed-bytes
2. typed-bytes to interface
3. interface to untyped-bytes
4. untyped-bytes to typedObject

this is good, but I'd also like
5. typed-bytes to typedObject

Comment (benluddy) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go]: Good catch, added that and "untyped-bytes to interface".

**Review by 7795114391 — COMMENTED** [review]
77951 14391 namaste

**Pull request merged**